### PR TITLE
[Draft] Introduce one-infer to one-cmds

### PR DESCRIPTION
**Pull request by yunjayh**
This commit introduces `one-infer` to `one-cmds` which enables inference tool
of backend driver, and the other runtimes, including tflite(included) and onnx(not yet)

Signed-off-by: yunjay-hong <yunjay.hong@samsung.com>

---

for: #8970

### CODE REVIEW

File: compiler/one-cmds/one-infer [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    if len(args) and infer_args.version:
+        _utils._print_version_and_exit(__file__)
+
+    return infer_args, backend_args, unknown_args
+
+
+def _get_infer_path(runtime):
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    # TODO: Add onnx runtime
+    available_runtime = ['tflite-infer']
```
Comment (seanshpark) [inline compiler/one-cmds/one-infer:134]: why is `flite-infer` added as fixed item?
can you explain or share any links for this?

Comment (seanshpark) [reply compiler/one-cmds/one-infer:134]: below caller for _get_infer_path is commented out...

Comment (yunjayh) [reply compiler/one-cmds/one-infer:134]: my bad.. It's deprecated caller which defined while drafting. I'll remove this. Thank you for noticing me.

**Comment by seanshpark** [comment]
can you explain how the usage of `tflite-infer` would be?
it's installed in `bin` but I think there seems to be one of the target called by `one-infer`...

**Comment by yunjayh** [comment]
`one-infer` searches the executable which matches with `*-infer` from both `one/backends/**` and `one/bin/`. 
if user runs `onecc infer -b tflite -- [options to tflite-infer]`, then `tflite-infer [options ...]` will be executed. 

Did I get your intention?

**Comment by seanshpark** [comment]
> Did I get your intention?

Yes, I got the idea :)

File: compiler/one-cmds/dummy-driver/src/dummy-infer.cpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * dummy-infer only tests its interface rather than its functionality.
```
Comment (YongseopKim) [inline compiler/one-cmds/dummy-driver/src/dummy-infer.cpp:18]: If so, `fstream` and `string` are unnecessary, I think.

File: compiler/one-cmds/dummy-driver/src/help-infer.cpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+int main(int argc, char **argv)
+{
+  if (argc != 2)
+    return EXIT_FAILURE;
+
+  std::string opt_h{"-h"};
+  std::string argv_1{argv[1]};
+
+  if (opt_h != argv_1)
```
Comment (YongseopKim) [inline compiler/one-cmds/dummy-driver/src/help-infer.cpp:36]: If so,
```c++
if (argc != 2 or string("-h") != string(argv[1]))
  return;
```

File: compiler/one-cmds/one-infer [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    # backends folder
+    files += [
+        f for f in glob.glob(dir_path + '/../backends/**/*-infer', recursive=True)
+    ]
+    # TODO find backends in `$PATH`
+
+    backends_list = []
+    for cand in files:
+        base = ntpath.basename(cand)
+        if not base in backend_set and os.path.isfile(cand) and os.access(cand, os.X_OK):
```
Comment (YongseopKim) [inline compiler/one-cmds/one-infer:68]: Do I understand right?

```py
if not base in backend_set and os.path.isfile(cand) and os.access(cand, os.X_OK):
```

means like below?
```py
if (not base in backend_set) and (os.path.isfile(cand)) and (os.access(cand, os.X_OK)):
```

and is it okay if it is symbolic link? I don't know how to handle it by `os.path.isfile()`.

File: compiler/one-cmds/tests/one-infer_001.test [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+filename_ext="$(basename -- $0)"
+filename="${filename_ext%.*}"
+
+trap_err_onexit()
```
Comment (YongseopKim) [inline compiler/one-cmds/tests/one-infer_001.test:20]: In the future, `trap_err_onexit` can be extracted to one of common functions like in one-cmds/tests/commoon.sh.

File: compiler/one-cmds/tflite-infer [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    return args
+
+
+def get_input_data(shape, dtype, cond='any', idx=0, filename=None):
+    cond_list = ['any', 'non-zero', 'positive', 'npy']
+    if cond not in cond_list:
+        raise Exception("  Input Condition Error!")
+
+    # If load npy file to input tensor
+    if cond == 'npy':
```
Comment (YongseopKim) [inline compiler/one-cmds/tflite-infer:94]: below code can be refactored in the future with something like map.

**Comment by YongseopKim** [comment]
LGTM

File: compiler/one-cmds/one-infer (hunk 4 above)
Comment (yunjayh) [reply compiler/one-cmds/one-infer:68]: > means like below?

Yes! wrapping with parenthesis looks better. Thank you!

> and is it okay if it is symbolic link? I don't know how to handle it by `os.path.isfile()`.

In my environment, `os.path.isfile()` returns true for symlink. 😃

**Comment by yunjayh** [comment]
I think it needs to add test for `tflite-infer` to release.
-> Use `inception_v3.tflite`? (It's included in `prepare_test_materials.sh`)
-> If it's too big size, make simple tflite model while preparing test env?
----> this can make some problem if package version has been changed.

**Comment by yunjayh** [comment]
From now on, I'll make PRs about `one-infer` itself and add test codes for `tflite-infer` in this draft simultaneously. 

TODOs

- Introduce `one-infer`
  - [ ] Introduce `one-infer` code and add to cmakelist
  - [ ] Add tests including `dummy-infer` & `help-infer`
  - [ ] Add `one-infer` to debian package
  - [ ] Add manpage for `one-infer`
- Introduce `tflite-infer`
  - [ ] ...

**Comment by seanshpark** [comment]
> Add one-infer to debian package

this does not seem to be in this draft.

**Comment by yunjayh** [comment]
> this does not seem to be in this draft. planning to add in near future?

Oh, I was working on my local draft. I synced!

**Comment by yunjayh** [comment]
Here is updated manpage for `one-infer`. 
image

**Comment by yunjayh** [comment]
> @yunjayh , `backend name to use -d DRIVER, ...` looks strange... what does it mean?

OMG there was an typo in the source. I updated the screen capture image. PTAL

**Comment by yunjayh** [comment]
+) `-b` and `-d` options are mutually exclusive. That should be added.

**Comment by yunjayh** [comment]
@seanshpark I've updated the image PTAL

**Pull request closed without merging**